### PR TITLE
[tests] Test that nodes respond to getdata with notfound

### PR DESCRIPTION
**Pull request by maflcko**
If a node has not announced a tx at all, then it should respond to
getdata messages for that tx with notfound, to avoid leaking tx
origination privacy.

In the future this could be adjusted such that a node responds with
notfound when a tx has not been announced to us, but that seems
to be a more involved change. See e.g.
https://github.com/jnewbery/bitcoin/commits/pr14220.1

### CODE REVIEW

**Comment by jnewbery** [comment]
Concept ACK. Great test.

File: test/functional/test_framework/messages.py [hunk 1]
```diff
@@ -1232,6 +1232,26 @@ def __repr__(self):
...
 
+class msg_notfound:
+    __slots__ = ("vec")
+    command = b"notfound"
+
+    def __init__(self, vec=None):
+        if vec is None:
+            self.vec = []
+        else:
+            self.vec = vec
```
Comment (jamesob) [inline test/functional/test_framework/messages.py]: nit: more concise (if not exactly equivalent) as `self.vec = vec or []`

File: test/functional/p2p_leak_tx.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+class P2PLeakTxTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        gen_node = self.nodes[0]  # The block and tx generating node
+        gen_node.generate(1)
```
Comment (sanket1729) [inline test/functional/p2p_leak_tx.py]: Should we use `generate` RPC in tests at all after #14299 ?

File: test/functional/p2p_leak_tx.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        self.sync_all()
```
Comment (sanket1729) [inline test/functional/p2p_leak_tx.py]: I am new here, but is `sync_all()` necessary when there is only 1 node?

File: test/functional/test_framework/messages.py (hunk 1 above)
Comment (jnewbery) [reply test/functional/test_framework/messages.py]: I think Marco's version is clearer.

File: test/functional/p2p_leak_tx.py (hunk 2 above)
Comment (jnewbery) [reply test/functional/p2p_leak_tx.py]: `generate()` is now a method in `TestNode()` that aliases to the `generatetoaddress` RPC

File: test/functional/p2p_leak_tx.py (hunk 3 above)
Comment (jnewbery) [reply test/functional/p2p_leak_tx.py]: Entirely correct! This can be removed

File: test/functional/p2p_leak_tx.py (hunk 2 above)
Comment (sanket1729) [reply test/functional/p2p_leak_tx.py]: Understood. That was a neat way to do things .

File: test/functional/test_framework/messages.py [hunk 4]
```diff
@@ -1232,6 +1232,23 @@ def __repr__(self):
         return "msg_mempool()"
 
 
+class msg_notfound:
+    __slots__ = ("vec")
```
Comment (practicalswift) [inline test/functional/test_framework/messages.py]: This should be `__slots__ = ("vec",)` to force `tuple` instead of `str` :-)

Comment (MarcoFalke) [reply test/functional/test_framework/messages.py]: Ugh, I always get this wrong. Thx

**Comment by jnewbery** [comment]
Tested ACK fa78a2fc670d7162d6ba6d432188da6d6e5288ac

**Comment by sanket1729** [comment]
Tested ACK fa78a2fc670d7162d6ba6d432188da6d6e5288ac

**Pull request merged**